### PR TITLE
[Neuron] Adding support for adding/ overriding neuron configuration a…

### PR DESCRIPTION
**Pull request by hbikki**
…nd adding support for neuron model quantization configuration.

1) Quantization : Added support for int8 quantization for the neuron models, this is achieved by passing the `quant` config as a param tot he neuron config, the `NEURON_QUANT_DTYPE` helps us customize the dtype for the quantization.

2) Neuron Config: Update the default neuron config to support the latest neuron release's optimizations out of the box.

3) Dynamic Config update/Override : Provided support to override or pass neuron config from the engine args -> ModelConfig -> NeuronConfig , this will minimize the  updates to  vllm and provides more customizable configuration for clients.
 

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Review by liangfu — COMMENTED** [review]
Thanks @hbikki for contribuing. The proposed change looks good overall. I left some comments.

File: vllm/model_executor/layers/quantization/neuron_quant.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+class NeuronQuantConfig(QuantizationConfig):
+    """Int8 Quantization Config class for Neuron Backend."""
+
+    def __init__(
+        self,
+        dequant_dtype: str = "f16",
+        quantize_method: str = "vector_dynamic",
+    ) -> None:
+        self.quant_dtype = os.getenv("NEURON_QUANT_DTYPE", "s8")
```
[the remark this answers is not in the document]
Comment (hbikki) [reply vllm/model_executor/layers/quantization/neuron_quant.py:20]: the datatype we use for quantization in neuron doesn't match the torch d types or existing quant , so created a new quant config to support it.

File: vllm/model_executor/layers/quantization/neuron_quant.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        return cls(dequant_dtype=dequant_dtype,
+                   quantize_method=quantize_method)
+
+    def get_quant_method(self, layer: Module, prefix: str) -> Optional[Any]:
+        return self.get_tnx_quantization_config()
+
+    def get_scaled_act_names(self) -> List[str]:
+        return []
+
+    def get_tnx_quantization_config(self):
```
[the remark this answers is not in the document]
Comment (hbikki) [reply vllm/model_executor/layers/quantization/neuron_quant.py:57]: got it , as  the quant config is specific to tnx, I can do a tnx package check before using the config , any thoughts ?

File: vllm/model_executor/model_loader/neuron.py [hunk 3]
```diff
@@ -119,19 +136,56 @@ def _get_buckets(env: str, default_value: List[int]) -> List[int]:
...
+        continuous_batching_config
+    }
+    return default_neuron_args
+
+
+def _get_neuron_config_after_override(default_neuron_config,
+                                      overridden_neuron_config):
+    from transformers_neuronx.config import NeuronConfig
+    overridden_neuron_config = overridden_neuron_config or {}
+    combined_config = {**default_neuron_config, **overridden_neuron_config}
```
[the remark this answers is not in the document]
Comment (hbikki) [reply vllm/model_executor/model_loader/neuron.py:173]: The overridden config replaces the default neuron config

[the remark this answers is not in the document]
Comment (hbikki) [reply vllm/model_executor/model_loader/neuron.py:173]: example : when use use this as override config:

```
override_neuron_config={"cast_logits_dtype":"bfloat16", "collectives_layout":"Dummy"}
```

we get this result :
```
(Pdb) combined_config
{'collectives_layout': 'Dummy', 'attention_layout': 'BSH', 'fuse_qkv': True, 'quant': <transformers_neuronx.config.QuantizationConfig object at 0x7f259835a2f0>, 'continuous_batching': <transformers_neuronx.config.ContinuousBatchingConfig object at 0x7f259831d750>, 'weight_tiling': True, 'cast_logits_dtype': 'bfloat16'}
```

File: vllm/model_executor/model_loader/neuron.py [hunk 4]
```diff
@@ -119,19 +136,56 @@ def _get_buckets(env: str, default_value: List[int]) -> List[int]:
...
+    continuous_batching_config = ContinuousBatchingConfig(
+        batch_size_for_shared_caches=scheduler_config.max_num_seqs)
+    quant_config = {
+        "dequant_dtype": TORCH_DTYPE_TO_NEURON_AMP[model_config.dtype],
+        "quantize_method": "vector_dynamic"
+    }
+    neuron_quantization_config_builder = lambda quant: get_quantization_config(
+        quant).from_config(quant_config).get_quant_method(None, "")
+    # TODO: Add Paged attention config to the default neuron arguments.
+    default_neuron_args = {
```
[the remark this answers is not in the document]
Comment (hbikki) [reply vllm/model_executor/model_loader/neuron.py]: updated for better readability, thanks!

File: vllm/model_executor/model_loader/neuron.py [hunk 5]
```diff
@@ -119,19 +136,56 @@ def _get_buckets(env: str, default_value: List[int]) -> List[int]:
...
+        "collectives_layout":
+        "BSH",
+        "attention_layout":
+        "BSH",
+        "fuse_qkv":
+        True,
+        "quant":
+        neuron_quantization_config_builder(model_config.quantization)
+        if model_config.quantization else None,
```
[the remark this answers is not in the document]
Comment (hbikki) [reply vllm/model_executor/model_loader/neuron.py]: yep added weight_tiling, thanks!

File: vllm/model_executor/model_loader/neuron.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (liangfu) [reply vllm/model_executor/model_loader/neuron.py:168]: what happens if there is an overlap in the keyword, like:
```
override_neuron_config={"collectives_layout":"BSH", "collectives_layout":"HSB"}
```
?

File: vllm/model_executor/model_loader/neuron.py [hunk 6]
```diff
@@ -119,19 +136,51 @@ def _get_buckets(env: str, default_value: List[int]) -> List[int]:
...
+        quant).from_config(quant_config).get_quant_method(None, "")
+    # TODO: Add Paged attention config to the default neuron arguments.
+    default_neuron_args = dict(
+        collectives_layout=LAYOUT_BSH,
+        attention_layout=LAYOUT_BSH,
+        fuse_qkv=True,
+        quant=neuron_quantization_config_builder(model_config.quantization)
+        if model_config.quantization else None,
+        continuous_batching=continuous_batching_config,
+        weight_tiling=True)
```
Comment (liangfu) [inline vllm/model_executor/model_loader/neuron.py:160]: we might not want weight_tiling enabled, when weight quantization is not enabled.
pair weight_tiling with weight quantization config ?

File: vllm/model_executor/model_loader/neuron.py [hunk 7]
```diff
@@ -119,19 +136,51 @@ def _get_buckets(env: str, default_value: List[int]) -> List[int]:
     return buckets_list
 
 
+def _get_default_tnx_neuron_config(model_config: ModelConfig,
```
Comment (liangfu) [inline vllm/model_executor/model_loader/neuron.py:139]: also avoid `tnx` terminology here?

File: vllm/model_executor/model_loader/neuron.py (hunk 6 above)
Comment (hbikki) [reply vllm/model_executor/model_loader/neuron.py]: Sure, I couldn't find an env variable to enable it dynamically , so paired it with qunat

File: vllm/model_executor/model_loader/neuron.py (hunk 7 above)
Comment (hbikki) [reply vllm/model_executor/model_loader/neuron.py:139]: Done

File: vllm/model_executor/model_loader/neuron.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hbikki) [reply vllm/model_executor/model_loader/neuron.py]: no the dict supports only one key with the name , the second collectives_layout will overwrite the first one ,
Based on our offline sync update to `.update()` for readability.

File: vllm/model_executor/layers/quantization/neuron_quant.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (liangfu) [reply vllm/model_executor/layers/quantization/neuron_quant.py:21]: `s8` can be translated to `torch.int8`, and  `f8e4m3fn` and be translated to `torch.float8_e4m3fn`.
can we map the relation between neuron dtype and torch dtype?

**Review by liangfu — APPROVED** [review]
Thanks @hbikki for the update. It looks good.

I think we can leave the dtype mapping between neuron-specific and torch native data types as a follow-up.

**Review by simon-mo — APPROVED** [review]
